nemo/collections/tts/models/hifigan.py

```diff
@@ -0,0 +1,238 @@
+        """
+        Runs the generator, for inputs and outputs see input_types, and output_types
+        """
+        return self.generator(x=spec)
+
+    @typecheck(output_types={"audio": NeuralType(('B', 'T'), AudioSignal())})
+    def convert_spectrogram_to_audio(self, spec: 'torch.tensor') -> 'torch.tensor':
+        return self(spec=spec).squeeze(1)
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        audio, audio_len = batch
+        audio_mel, dummy = self.audio_to_melspec_precessor(audio, audio_len)
+        audio = audio.unsqueeze(1)
+
+        audio_pred = self.generator(x=audio_mel)
+        audio_pred_mel, _ = self.trg_melspec_fn(audio_pred.squeeze(1), audio_len)
+
+        # train discriminator
+        self.optim_d.zero_grad()
+        mpd_score_real, mpd_score_gen, _, _ = self.mpd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_mpd, _, _ = discriminator_loss(mpd_score_real, mpd_score_gen)
+        msd_score_real, msd_score_gen, _, _ = self.msd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_msd, _, _ = discriminator_loss(msd_score_real, msd_score_gen)
+        loss_d = loss_disc_msd + loss_disc_mpd
+        self.manual_backward(loss_d, self.optim_d)
+        self.optim_d.step()
+
+        # train generator
+        self.optim_g.zero_grad()
+        loss_mel = F.l1_loss(audio_pred_mel, audio_mel) * 45
+        _, mpd_score_gen, fmap_mpd_real, fmap_mpd_gen = self.mpd(y=audio, y_hat=audio_pred)
+        _, msd_score_gen, fmap_msd_real, fmap_msd_gen = self.msd(y=audio, y_hat=audio_pred)
+        loss_fm_mpd = feature_loss(fmap_mpd_real, fmap_mpd_gen)
+        loss_fm_msd = feature_loss(fmap_msd_real, fmap_msd_gen)
+        loss_gen_mpd, _ = generator_loss(mpd_score_gen)
+        loss_gen_msd, _ = generator_loss(msd_score_gen)
+        loss_g = loss_gen_msd + loss_gen_mpd + loss_fm_msd + loss_fm_mpd + loss_mel
+        self.manual_backward(loss_g, self.optim_g)
+        self.optim_g.step()
+
+        metrics = {
+            "g_l1_loss": loss_mel,
+            "g_loss_fm_mpd": loss_fm_mpd,
+            "g_loss_fm_msd": loss_fm_msd,
+            "g_loss_gen_mpd": loss_gen_mpd,
+            "g_loss_gen_msd": loss_gen_msd,
+            "g_loss": loss_g,
+            "d_loss_mpd": loss_disc_mpd,
+            "d_loss_msd": loss_disc_msd,
+            "d_loss": loss_d,
+            "global_step": self.global_step,
+            "lr": self.optim_g.param_groups[0]['lr'],
+        }
+        self.log_dict(metrics, on_step=False, on_epoch=True, sync_dist=True)
+
+    def validation_step(self, batch, batch_idx):
+        audio, audio_len = batch
+        audio_mel, audio_mel_len = self.audio_to_melspec_precessor(audio, audio_len)
+        audio_pred = self(spec=audio_mel)
+
+        audio_pred_mel, _ = self.audio_to_melspec_precessor(audio_pred.squeeze(1), audio_len)
+        loss_mel = F.l1_loss(audio_mel, audio_pred_mel)
+
+        self.log("val_loss", loss_mel, prog_bar=True, sync_dist=True)
```

Comment:
Doesn't this just log the loss at one step? As opposed to the validation loss over the entire val dataset.

Reply:
no, according to PTL, when you log in `val_step` the default behavior is to aggregate all steps into an epoch summary of the metric. see [here](https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#log).